pkg/apis/discovery/validation/validation_test.go

```diff
@@ -0,0 +1,329 @@
+*/
+
+package validation
+
+import (
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	api "k8s.io/kubernetes/pkg/apis/core"
+	"k8s.io/kubernetes/pkg/apis/discovery"
+)
+
+func TestValidateEndpointSlice(t *testing.T) {
+	standardMeta := metav1.ObjectMeta{
+		Name:      "hello",
+		Namespace: "world",
+	}
+
+	nameHTTP := "http"
+	nameEmpty := ""
+	nameCaps := "ABCDEF"
+	nameChars := "almost_valid"
+	hostnameInvalid := "--invalid"
+	hostnameValid := "valid-123"
+	protoTCP := api.ProtocolTCP
+	addressTypeIP := discovery.AddressTypeIP
+	addressTypeOther := discovery.AddressType("Other")
+
+	testCases := map[string]struct {
+		isExpectedFailure bool
+		endpointSlice     *discovery.EndpointSlice
+	}{
+		"good-slice": {
+			isExpectedFailure: false,
+			endpointSlice: &discovery.EndpointSlice{
+				ObjectMeta:  standardMeta,
+				AddressType: &addressTypeIP,
+				Ports: []discovery.EndpointPort{{
+					Name:     &nameHTTP,
+					Protocol: &protoTCP,
+				}},
+				Endpoints: []discovery.Endpoint{{
+					Addresses: []string{"1.2.3.4"},
+					Hostname:  &hostnameValid,
+				}},
+			},
+		},
+		"empty-port-name": {
+			isExpectedFailure: false,
+			endpointSlice: &discovery.EndpointSlice{
+				ObjectMeta:  standardMeta,
+				AddressType: &addressTypeIP,
+				Ports: []discovery.EndpointPort{{
+					Name:     &nameEmpty,
+					Protocol: &protoTCP,
+				}, {
+					Name:     &nameHTTP,
+					Protocol: &protoTCP,
+				}},
+				Endpoints: []discovery.Endpoint{{
+					Addresses: []string{"1.2.3.4"},
+				}},
+			},
+		},
+		"empty-ports-and-endpoints": {
+			isExpectedFailure: false,
+			endpointSlice: &discovery.EndpointSlice{
+				ObjectMeta:  standardMeta,
+				AddressType: &addressTypeIP,
+				Ports:       []discovery.EndpointPort{},
+				Endpoints:   []discovery.Endpoint{},
+			},
+		},
+		"max-topology-keys": {
+			isExpectedFailure: false,
+			endpointSlice: &discovery.EndpointSlice{
+				ObjectMeta:  standardMeta,
+				AddressType: &addressTypeIP,
+				Ports: []discovery.EndpointPort{{
+					Name:     &nameHTTP,
+					Protocol: &protoTCP,
+				}},
+				Endpoints: []discovery.Endpoint{{
+					Addresses: []string{"1.2.3.4"},
+					Topology: map[string]string{
+						"one":      "example",
+						"two":      "example",
+						"three":    "example",
+						"four":     "example",
+						"five":     "example",
+						"six":      "example",
+						"seven":    "example",
+						"eight":    "example",
+						"nine":     "example",
+						"ten":      "example",
+						"eleven":   "example",
+						"twelve":   "example",
+						"thirteen": "example",
+						"fourteen": "example",
+						"fifteen":  "example",
+						"sixteen":  "example",
+					},
+				}},
+			},
+		},
+
+		// expected failures
+		"duplicate-port-name": {
+			isExpectedFailure: true,
+			endpointSlice: &discovery.EndpointSlice{
+				ObjectMeta:  standardMeta,
+				AddressType: &addressTypeIP,
+				Ports: []discovery.EndpointPort{{
+					Name:     &nameEmpty,
```

Comment:
IMO, this test would be more readable with some helper functions, e.g. `newString(v string) { return &v }` or even `makePort(name string, proto string, num int) { return discovery.EndpointPort{ ... } }`

This is especially true for the max-len tests.  `makeSSMap(16) map[string]string` and then internally iterate and produce keys like "x1", "x2", x3"...